src/Sentry.Unity/UnityEventProcessor.cs

```diff
@@ -30,104 +28,119 @@ public UnityEventProcessor(IApplication? application = null)
 
         public SentryEvent Process(SentryEvent @event)
         {
-            @event.Sdk.AddPackage(UnitySdkInfo.PackageName, UnitySdkInfo.Version);
-            @event.Sdk.Name = UnitySdkInfo.Name;
-            @event.Sdk.Version = UnitySdkInfo.Version;
-
-            @event.Contexts.OperatingSystem.Name = SystemInfo.operatingSystem;
-
-            @event.Contexts.Device.Name = SystemInfo.deviceName;
-#pragma warning disable RECS0018 // Value is exact when expressing no battery level
-            if (SystemInfo.batteryLevel != -1.0)
-#pragma warning restore RECS0018
-            {
-                @event.Contexts.Device.BatteryLevel = (short?)(SystemInfo.batteryLevel * 100);
-            }
+            PopulateSdk(@event.Sdk);
+            PopulateApp(@event.Contexts.App);
+            PopulateOperatingSystem(@event.Contexts.OperatingSystem);
+            PopulateDevice(@event.Contexts.Device, _application);
+            PopulateGpu(@event.Contexts.Gpu);
+            PopulateUnity((Protocol.Unity)@event.Contexts.GetOrAdd(Protocol.Unity.Type, _ => new Protocol.Unity()));
 
             @event.ServerName = null;
 
-            // This is the approximate amount of system memory in megabytes.
-            // This function is not supported on Windows Store Apps and will always return 0.
-            @event.Contexts.Device.MemorySize = SystemInfo.systemMemorySize;
+            return @event;
+        }
 
-            @event.Contexts.Device.Timezone = TimeZoneInfo.Local;
+        private static void PopulateSdk(SdkVersion sdk)
+        {
+            sdk.AddPackage(UnitySdkInfo.PackageName, UnitySdkInfo.Version);
+            sdk.Name = UnitySdkInfo.Name;
+            sdk.Version = UnitySdkInfo.Version;
+        }
 
-            @event.Contexts.App.StartTime = DateTimeOffset.UtcNow.AddSeconds(-Time.realtimeSinceStartup);
+        private static void PopulateApp(App app)
+        {
+            app.StartTime = DateTimeOffset.UtcNow
+                // NOTE: Time API requires main thread
```

Comment:
Can we do a runtime check? We better not run this if we're not on the main thread.
Fair point this is unrelated to the PR but rather broken right now

Reply:
Check if we're on the main thread before doing that?